File: test/executable/test_exe.f90
program test_exe

use, intrinsic :: iso_fortran_env, only : stderr=>error_unit
use filesystem

implicit none


call test_not_exist()
print '(a)', "PASSED: test_not_exist"

call test_is_exe()
print '(a)', "PASSED: test_exist"

contains

subroutine test_not_exist()

character(:), allocatable :: s1

!> empty file
if(is_file("")) error stop "ERROR:test_exe: is_file('') should be false"
if(is_exe("")) error stop "ERROR:test_exe: is_exe('') should be false"

s1 = get_permissions("")
if(len_trim(s1) /= 0) error stop "ERROR:test_exe: get_permissions('') should be empty: " // s1

!> not exist file
s1 = "not-exist-file"
if (is_file(s1)) error stop "ERROR:test_exe: not-exist-file should not exist."
if (is_exe(s1)) error stop "ERROR:test_exe: non-exist-file cannot be executable"
if(len_trim(get_permissions(s1)) /= 0) error stop "ERROR:test_exe: get_permissions('not-exist') should be empty"

end subroutine test_not_exist


subroutine test_is_exe()

character(*), parameter :: noexe = "test_noexe"

character(:), allocatable :: exe

if(is_wsl() > 0 .and. filesystem_type(".") == "v9fs") then
  print '(a)', "XFAIL:test_exe: WSL with VFS does not support permissions."
  stop 77
end if

exe = getarg(0)
if(len_trim(exe) == 0 ) error stop "could not get own program name"

print '(a)', "test_is_exe: touch(" // noexe // ")"
call touch(noexe)

print '(a)', "set_permissions(" // trim(noexe) // ", executable=.false.)"
call set_permissions(noexe, executable=.false.)

if(is_exe(parent(exe))) then
  write(stderr, '(a)') "ERROR:test_exe: directory" // parent(exe) // " should not be executable"
  error stop
end if

print '(a)', "permissions: " // trim(exe) // " = " // get_permissions(exe)

if (.not. is_exe(exe)) then
  write(stderr,'(a)') "ERROR:test_exe: " // trim(exe) // " is not executable."
  error stop
end if

print '(a)', "permissions: " // trim(noexe) // " = " // get_permissions(noexe)

if (is_exe(noexe)) then
  if(is_windows()) then
    write(stderr,'(a)') "XFAIL:test_exe: " // trim(noexe) // " is executable on Windows."
  else
    write(stderr,'(a)') "ERROR:test_exe: " // trim(noexe) // " is executable and should not be."
    error stop
  end if
end if

print '(a)', "remove(" // trim(noexe) // ")"
call remove(noexe)

end subroutine test_is_exe

end program

File: test/executable/test_exe.cpp
#include <iostream>
#include <cstdlib>
#include <string>

#include "ffilesystem.h"

#ifdef _MSC_VER
#include <crtdbg.h>
#endif

#include "ffilesystem_test.h"


int main()
{

#ifdef _MSC_VER
  _CrtSetReportMode(_CRT_ASSERT, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_ASSERT, _CRTDBG_FILE_STDERR);
  _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
  _CrtSetReportMode(_CRT_ERROR, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_ERROR, _CRTDBG_FILE_STDERR);
#endif

auto cwd = fs_get_cwd();
if(cwd.empty())
  return EXIT_FAILURE;

if(fs_is_wsl() && fs_filesystem_type(cwd) == "v9fs"){
  std::cerr << "XFAIL:WSL: test_exe: v9fs to NTFS etc. doesn't work right\n";
  std::exit(77);
}

std::string exe = "test_exe";
std::string noexe = "test_noexe";

// Empty string
if(fs_is_exe(""))
    err("test_exe: is_exe('') should be false");

// Non-existent file
if (fs_is_file("not-exist"))
    err("test_exe: not-exist-file should not exist.");
if (fs_is_exe("not-exist"))
    err("test_exe: not-exist-file cannot be executable");

fs_touch(exe);
fs_touch(noexe);

fs_set_permissions(exe, 0, 0, 1);
fs_set_permissions(noexe, 0, 0, -1);

if(fs_is_exe(fs_parent(exe)))
    err("test_exe: is_exe() should not detect directory " + fs_parent(exe));

std::string p = fs_get_permissions(exe);
if(p.empty())
  err("test_exe: get_permissions(" + exe + ") failed");

std::cout << "permissions: " << exe << " = " << p << "\n";

if (!fs_is_exe(exe))
  err("test_exe: " + exe + " is not executable and should be.");

p = fs_get_permissions(noexe);
if(p.empty())
  err("test_exe: get_permissions(" + noexe + ") failed");
std::cout << "permissions: " << noexe << " = " << p << "\n";

if (fs_is_exe(noexe)){
  if(fs_is_windows()){
    std::cerr << "XFAIL:Windows: test_exe: is_exe() did not detect non-executable file " << noexe << " on Windows\n";
  }
  else{
   err("test_exe: " + noexe + " is executable and should not be.");
  }
}

// chmod setup

fs_remove(exe);
fs_remove(noexe);

// chmod(true)
fs_touch(exe);
if (!fs_is_file(exe))
    err("test_exe: " + exe + " is not a file.");

p = fs_get_permissions(exe);
if(p.empty())
  err("test_exe: get_permissions(" + exe + ") failed");

std::cout << "permissions before chmod(" << exe << ", true)  = " << p << "\n";

fs_set_permissions(exe, 0, 0, 1);

p = fs_get_permissions(exe);
if(p.empty())
  err("test_exe: get_permissions(" + exe + ") failed after set");
std::cout << "permissions after chmod(" << exe << ", true) = " << p << "\n";

if (!fs_is_exe(exe)){
  if(fs_is_windows()){
    std::cerr << "XFAIL:Windows: test_exe: is_exe() did not detect executable file " << exe << " on Windows\n";
  }
  else{
    err("test_exe: is_exe() did not detect executable file " + exe);
  }
}

if (!fs_is_windows() && p[2] != 'x'){
  err("test_exe: expected POSIX perms for " + exe + " to be 'x' in index 2");
}

// chmod(false)
fs_touch(noexe);
if (!fs_is_file(noexe))
    err("test_exe: " + noexe + " is not a file.");

p = fs_get_permissions(noexe);
if(p.empty())
  err("test_exe: get_permissions(" + noexe + ") failed after touch");
std::cout << "permissions before chmod(" << noexe << ", false)  = " << p << "\n";

fs_set_permissions(noexe, 0, 0, 0);

p = fs_get_permissions(noexe);
if(p.empty())
  err("test_exe: get_permissions(" + noexe + ") failed after set");

std::cout << "permissions after chmod(" << noexe << ",false) = " << p << "\n";

if(!fs_is_windows())
{
  if (fs_is_exe(noexe))
    err("test_exe: did not detect non-executable file.");

  if (p[2] != '-')
    err("test_exe: expected POSIX perms for " + noexe + " to be '-' in index 2");
}

// test which
std::string_view name = (fs_is_windows()) ? "cmake.exe" : "ls";

std::string r = fs_which(name);
if(r.length() == 0)
  std::cerr << "ERROR:test_exe: which(" << name << ") should return a path. This can happen on CI systems.\n";

std::cout << "which(" << name << ") " << r << "\n";

// our own chmod(exe)
r = fs_which(exe);
if(r.length() == 0)
  err("test_exe: which(" + exe + ") should return a path.");

std::cout << "which(" << exe << ") " << r << "\n";


fs_remove(exe);
fs_remove(noexe);

std::cout << "OK: c++ test_exe\n";

return EXIT_SUCCESS;
}
